blocks/graph.py

```diff
@@ -161,8 +162,40 @@ def replace(self, replacements):
             substitutes.
 
         """
-        return ComputationGraph(theano.clone(self.outputs,
-                                             replace=replacements))
+        # Due to theano specifics we have to make one replacement in time
+        replacements = OrderedDict(replacements)
+
+        apply_nodes_sorted = theano.gof.graph.io_toposort(
+            self.inputs, self.outputs)
+        outputs_cur = self.outputs
+
+        # `replacements` with previous replacements applied. We have to track
+        # variables in the new graph corresponding to original replacements.
+        replacements_cur = OrderedDict()
+        # Sort `replacements` in topological order
+        for node in apply_nodes_sorted:
+            for input_ in node.inputs:
+                if input_ not in replacements:
+                    continue
+                replacements_cur[input_] = replacements[input_]
+
+        # Replace step-by-step in topological order
+        for i in range(len(replacements_cur)):
```

Comment:
Yes, I see that now. Although it seems like it'd be easier to just do 
```python
keys, values = replacements_cur.keys(), replacements_cur.values()
```
only once before the loop, and replace the current `replacement = ` line with
```python
replacement = (keys[i], values[i])
```
Then you can operate on the key and value lists directly, since you don't seem to need it to be a dictionary within the loop.